client/clientservice/src/event_service.cpp

```diff
@@ -119,11 +127,28 @@ Status EventServiceImpl::Subscribe(ServerContext* context,
       // TODO: Set trace context
 
       *response.mutable_events() = proto_events;
+      std::chrono::steady_clock::time_point write_start = std::chrono::steady_clock::now();
       stream->Write(response);
-
+      num_writes++;
+      // update write duration metric
+      end_overall = std::chrono::steady_clock::now();
+      auto duration_write = std::chrono::duration_cast<std::chrono::milliseconds>(end_overall - write_start);
+      metrics_.write_dur.Get().Set(duration_write.count());
     } else {
       LOG_ERROR(logger_, "Got unexpected update type from TRC. This should never happen!");
     }
+    // update overall duration metric
+    auto update_duration_overall = std::chrono::duration_cast<std::chrono::microseconds>(end_overall - start_overall);
+    metrics_.update_processing_dur.Get().Set(update_duration_overall.count());
+
+    // update writes/sec metric
+    auto duration_overall = std::chrono::duration_cast<std::chrono::seconds>(end_overall - start_writes_per_second);
+    if (duration_overall >= std::chrono::seconds(1)) {
+      metrics_.num_writes_per_second.Get().Set(num_writes / duration_overall.count());
+      num_writes = 0;
+      start_writes_per_second = std::chrono::steady_clock::now();
+    }
+    metrics_.updateAggregator();
```

Comment:
When we update the values inside the aggregator we acquire a lock. I wonder if this function might be called too frequently and therefore lead to unnecessary waiting. Wouldn't it be sufficient to call it once every second? Meaning, call it whenever we calculate writes per second. @yontyon Might have suggestions.

Reply:
Yes, the metrics are eventually consistent, so calling the update aggregator once in a while is fine.
Notice, however, that as the metrics themselves are not thread-safe, you must call the updateAggregator from the same thread in which you update the metrics.
Note that the aggregator lock is for mutual exclusion between the metrics server thread (e.g. Prometheus HTTP server) and this should happen only once in a few seconds. With modern implementation, grabbing mutex when there is no contention is fast as test&set (see https://preshing.com/20111118/locks-arent-slow-lock-contention-is/).

Reply:
Thanks @yontyon updateAggregator will be called by the same thread that would update the metrics.